c10/core/CPUCachingAllocator.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <algorithm>
+#include <deque>
+#include <memory>
+#include <mutex>
+
+#include <c10/core/CPUAllocator.h>
+#include <c10/util/Exception.h>
+#include <c10/util/flat_hash_map.h>
+
+namespace c10 {
+
+class CPUCachingAllocator {
+  /*
+   * What it does:
+   * Caches all the allocations carried out by this allocator.
+   * Cache key is the size of the allocation.
+   * If requested size is found in the cache returns the cached pointer.
+   * What it does not do:
+   * No speculative allocation for any future allocations.
+   */
+  private:
+    std::mutex mutex_;
+    // Invariants.
+    // 1. If an allocated memory pointer exists in allocation_map_ then
+    //    it must not exist in available_map_, and vice-e-versa.
+    // 2. All the pointers in available_map_ must be unique.
+    ska::flat_hash_map<void*, size_t> allocation_map_;
+    ska::flat_hash_map<size_t, std::deque<void*>> available_map_;
```

Comment:
IIRC, `std::deque` has pretty bad constant factor performance (though it might be better in libc++).  I would make this a vector (or even c10::SmallVector) and treat it as a stack.  FIFO also has the potential to give better temporal locality in the CPU cache.

Reply:
I see. I was avoiding vector to avoid dynamic sizing and copying on allocation path. Did not know about bad constant factor perf of deque though.

Reply:
Also another reason to use deque was that the only ops needed were push_back and pop_front or pop_back so I thought it will be faster with deque. But I think SmallVector should do as well.

Reply:
Currently SmallVector isn't compiling with flat_hash_map, @smessmer has a fix in https://github.com/pytorch/pytorch/pull/42694. Will rebase once that lands.